agent/hcp/telemetry_provider.go

```diff
@@ -322,7 +334,30 @@ func (h *hcpProviderImpl) setRunning(r bool) bool {
 		return false
 	}
 
+	// Initialize or close the stop channel depending what running status
+	// we're transitioning to. Channel must be initialized on start since
+	// a provider can be stopped and started multiple times.
+	if r {
+		h.stopCh = make(chan struct{})
+	} else {
+		close(h.stopCh)
+	}
+
 	h.running = r
 
 	return true
 }
+
+// Stop acquires a write lock to mark the provider as not running and sends a stop signal to the
+// main run loop. It also updates the provider with a disabled configuration.
+func (h *hcpProviderImpl) Stop() {
+	changed := h.setRunning(false)
+	if !changed {
+		h.logger.Trace("telemetry config provider already stopped")
+		return
+	}
+
+	h.cfg = defaultDisabledCfg()
```

Comment:
Do we need to lock before/after this? Or is this assignment atomic in Go? I'm not sure

Reply:
You're right, this needs to be locked!